### PR TITLE
sync: pull only missing images, not everything, closes #335

### LINKED ISSUE #335: sync: when in poll mode, download only what is missing, not everything
**Describe the bug**
Currently, in poll mode, sync downloads everything that is listed upstream and then tries to push to local repo which over time will already be populated with images, so we are just using up network. Use delta-sync - download only what is missing.

**Expected behavior**
Initial state - no local images, download everything
Steady state - SET_DIFF (upstream images  - local images) and only download those.

### PR DESCRIPTION
**Pull request by peusebiu**
Signed-off-by: Petu Eusebiu <peusebiu@cisco.com>

**What type of PR is this?**

**Which issue does this PR fix**:

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs from IPAMD/CNI showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades. Has updating a running cluster been tested?**:

**Does this change require updates to the CNI daemonset config files to work?**:

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/extensions/sync/utils.go [hunk 1]
```diff
@@ -155,3 +157,43 @@ func pushSyncedLocalImage(repo, tag, uuid string,
...
+// retrieve all local images.
+func getLocalImages(storeController storage.StoreController) (localImages, error) {
+	imageStoreList := make(map[string]storage.ImageStore)
+	if storeController.SubStore != nil {
+		imageStoreList = storeController.SubStore
+	}
+
+	imageStoreList[""] = storeController.DefaultStore
+
+	localImages := make(localImages)
```
Comment (adodon2go) [inline pkg/extensions/sync/utils.go:170]: confusing to read.

File: pkg/extensions/sync/utils.go [hunk 2]
```diff
@@ -155,3 +157,43 @@ func pushSyncedLocalImage(repo, tag, uuid string,
...
+			}
+
+			localImages[repo] = tags
+		}
+	}
+
+	return localImages, nil
+}
+
+func contains(slice []string, item string) bool {
```
[the remark this answers is not in the document]
Comment (peusebiu) [reply pkg/extensions/sync/utils.go:199]: do you have any suggestion? I can not import pkg/api package in sync due tot circular imports

File: pkg/extensions/sync/utils.go [hunk 3]
```diff
@@ -17,6 +17,8 @@ import (
 	"zotregistry.io/zot/pkg/storage"
 )
 
+type localImages map[string][]string
```
[the remark this answers is not in the document]
Comment (peusebiu) [reply pkg/extensions/sync/utils.go]: fixed, thank you!

File: pkg/extensions/sync/utils.go (hunk 1 above)
Comment (peusebiu) [reply pkg/extensions/sync/utils.go]: fixed

File: pkg/extensions/sync/utils.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (adodon2go) [reply pkg/extensions/sync/utils.go:199]: How about creating a new pkg like pkg/common.
We already have something similar in test/common.go but that is for testing purpose.
I'm pretty sure that we still have duplicate code in the project that can be potentially moved in a common location.

[the remark this answers is not in the document]
Comment (rchincha) [reply pkg/extensions/sync/utils.go:199]: pkg/test pls

[the remark this answers is not in the document]
Comment (rchincha) [reply pkg/extensions/sync/utils.go:199]: and move test/*.go there and test data under internal/test/

[the remark this answers is not in the document]
Comment (rchincha) [reply pkg/extensions/sync/utils.go:199]: If this pattern is in multiple places, yes, would prefer a pkg/utils or a pkg/common

**Review by rchincha — APPROVED** [review]
lgtm

**Pull request merged**